File: Go/LongestSubstr.go
package main

func lengthOfLongestSubstring(s string) int {
	n := len(s)
	ans, i, j := 0, 0, 0
	var set map[rune]bool
	set = make(map[rune]bool)

	for i < n && j < n {
		charAtJ := []rune(s)[j]
		if !set[charAtJ] {
			set[charAtJ] = true
			j++
			ans = Max(ans, j-i)
		} else {
			charAtI := []rune(s)[i]
			set[charAtI] = false
			i++
		}
	}

	return ans
}

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: Typescript/LongestSubstr.ts
export interface ISet {
    [props: string] : boolean;
}


function lengthOfLongestSubstring(s: string): number {
    let i=0,j=0,ans=0,n = s.length;
    let set: ISet = {};
    while(i<n && j<n){
        let curchar = s.charAt(j);
        if( !set.hasOwnProperty(curchar) ){
            set[curchar] = true;
            j++;
            ans = Math.max(ans, j-i);
        }else{
            let curcharI = s.charAt(i)
            delete set[curcharI];
            i++
        }
    }
    return ans;
};

console.log(lengthOfLongestSubstring("abba"));